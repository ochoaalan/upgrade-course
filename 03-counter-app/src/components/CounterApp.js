// import React, { Fragment } from 'react';
import React, { useState } from "react";
import PropTypes from "prop-types";

function CounterApp({ value = 10, subtitulo }) {
  const [counter, setCounter] = useState( value );

  // handleAdd
  const handleAdd = (e) => {
    setCounter( counter + 1 );
    // otra forma de usar setCounter
    // setCounter((contador) => contador + 1);
  };

    // handleSubtract
    const handleSubtract = (e) => {
      setCounter( counter - 1 );
    };

    // handleReset
    const handleReset = (e) => {
      setCounter( value );
    }

  return (
    <>
      <h1>Counter App</h1>
      <p>{ subtitulo }</p>
      <h2> {counter} </h2>

      {/* <button onClick={(e) => { handleAdd(e) }}>+1</button> */}
      {/* simplificando funcion */}
      <button onClick={ handleAdd }>+1</button>
      <button onClick={ handleReset }>Reset</button>
      <button onClick={ handleSubtract }>-1</button>
      {/* otra forma usando la funcion setcounter aqui directo */}
      {/* <button onClick={ ()=> setCounter( value ) }>Reset</button>
      <button onClick={ ()=> setCounter( counter -1) }>-1</button> */}
    </>
  );
}

  CounterApp.propTypes = {
// value: PropTypes.number.isRequired,
  value: PropTypes.number
  };

  CounterApp.defaultProps = {
  subtitulo: 'Soy un subtitulo'
  }

export default CounterApp;
