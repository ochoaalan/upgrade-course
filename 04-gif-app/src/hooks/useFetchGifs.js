import { useEffect, useState } from "react"
import { getGifs } from '../helpers/getGifs'


export const useFetchGifs = ( category ) => {

  const [state, setstate] = useState({
    data: [],
    loading: true
  });

  useEffect( () => { // para que solo se dispare cuando cambie la categoria

    getGifs( category ) // aqui hace la peticion http
    .then( imgs => { // obtenemos las imagenes en el result como imgs o cualquier nombre que quieras
        // setState dispara una re-renderizacion del componente GifGrid porque es el que nos llamo y es lo damos de rreturn
        setstate({ // Cambiamos la data con la recibida
          data: imgs,
          loading: false
        });
    })

  }, [category])

  return state; //{data:[], loading: true}

}