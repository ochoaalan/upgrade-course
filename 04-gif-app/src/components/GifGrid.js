import React from 'react';
import PropTypes from 'prop-types';

import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

  const { data:images, loading } = useFetchGifs( category ); // data:images  cambia el nombre de data a images en la desestrutcturacion para que tenga mas sentido nuestro codigo

  return (
    <>
      <h3 className='animate__animated animate__fadeIn'>{ category }</h3>

      {loading && <p className='animate__animated animate__flash'>Loading...</p>}

        <div className='card-grid'>
        {
          images.map( img => ( 
            <GifGridItem 
              key={ img.id }
              {...img }
            />
          ))
        }
        </div>
    </>

  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}