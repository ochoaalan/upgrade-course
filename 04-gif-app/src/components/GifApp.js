import React, { useState } from 'react';
import { AddCategory } from './AddCategory';
import { GifGrid } from './GifGrid';

export const GifApp = ({ defaultCategories = [] }) => {

// const [categories, setCategories] = useState(['Castlevania']);
const [categories, setCategories] = useState( defaultCategories );

  return (
    <div>
      <h2>GifApp</h2>
      <AddCategory setCategories={setCategories} />
      <hr />

      <ol>
        { 
          categories.map( category => (
            <GifGrid 
              key={ category }
              category={ category }
            />
          ))
        }
      </ol>
    </div>
  )
}
