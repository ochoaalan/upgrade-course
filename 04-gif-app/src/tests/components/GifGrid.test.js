import React from "react";
import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

jest.mock("../../hooks/useFetchGifs");

describe("Tests for GifGrid component", () => {
  const category = "Goku";
  // const wrapper = shallow(<GifGrid category={category}/>);

  test("should show the correct component and match with the snapshot", () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });
    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar items cuando se cargan imagenes useFetchGifs", () => {
    const gifs = [
      {
        id: "ABC",
        url: "https://google.com",
        title: "ABC Title",
      },
      {
        id: "123",
        url: "https://localhost/cualquier/cosa.jpg",
        title: "123 Cualquier cosa",
      },
    ];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
