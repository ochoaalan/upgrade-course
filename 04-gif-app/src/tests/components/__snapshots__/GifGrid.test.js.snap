// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for GifGrid component Debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Goku
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="ABC Title"
      url="https://google.com"
    />
    <GifGridItem
      id="123"
      key="123"
      title="123 Cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
  </div>
</Fragment>
`;

exports[`Tests for GifGrid component should show the correct component and match with the snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Goku
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
